Add validation to message constructors

harbu opened the pull request:
Add validation to the constructors of message classes. Two benefits. Firstly, allows us to catch errors caused by missing fields earlier in the execution of a request. Secondly, allows us to get rid of some basic validation checks in client and broker as we can assume that presence (and types) of fields have been checked in constructor. 

### Changeset
- 1st commit: validation functions and unit tests for them.
- 2nd commit: adding (and replacing existing) validation into messages classes. I'm not entirely sure about the optionality of some of the fields, especially those of the older versions V28, V29, and V0. Thus I opted to assume optionality in cases in which I was unsure. 
- 3rd commit: this is a change to the behavior of `toOtherVersion` which can potentially break something if we are depending on this behavior somewhere. It is quite arbitrary, so I'm wondering if this is the case really.
4th commit: Fix a few tests, e.g., some were using integers as msgChainId. Some fixes are related to the behavior change introduced in the 3rd commit.

### Version upgrade
This PR has a high chance of breaking something in broker or client. Thus we should treat it as a _major_ version upgrade and take necessary precautions before migrating broker or client to use this changeset.

mirpo commented on test/unit/protocol/message_layer/StreamMessageV28.test.js:
```diff
@@ -71,7 +71,7 @@ describe('StreamMessageV28', () => {
         it('correctly serializes messages to v31', () => {
             const arr = [31, ['TsvTbqshTsuLg_HyUjxigA', 0, 1529549961116, 0, '', ''],
-                [null, null], StreamMessage.CONTENT_TYPES.MESSAGE, StreamMessage.ENCRYPTION_TYPES.NONE, '{"valid": "json"}', 0, null]
+                null, StreamMessage.CONTENT_TYPES.MESSAGE, StreamMessage.ENCRYPTION_TYPES.NONE, '{"valid": "json"}', 0, null]
```
why it was [null, null] and now its just null? what if I'll put ['blabla', null]?

harbu replied:
First question: dunno. Second question: example will not pass validation.

mirpo commented on test/unit/protocol/message_layer/StreamMessageV30.test.js:
```diff
@@ -128,7 +128,7 @@ describe('StreamMessageV30', () => {
             }
             const msg = new StreamMessageV30(
-                ['streamId', 0, Date.now(), 0, 'publisherId', 1], [1529549961000, 0], StreamMessage.CONTENT_TYPES.MESSAGE, content,
+                ['streamId', 0, Date.now(), 0, 'publisherId', '1'], [1529549961000, 0], StreamMessage.CONTENT_TYPES.MESSAGE, content,
```
not int? 👍  :) https://github.com/streamr-dev/streamr-client-protocol-js/blob/master/PROTOCOL.md#messageid

harbu commented:
I tested this as follows.
- streamr-network: updated protocol first to `3.0.0-0` and then to `3.0.0-1` and in both situations ran tests. A [few fixes](https://github.com/streamr-dev/network/pull/361) are necessary for tests to pass but no major problems.
- streamr-client-javascript: updated protocol to `3.0.0-0`. Found a few fields that I thought were mandatory but were indeed optional. Updated these issues in protocol `3.0.0-1`. Now streamr-client-javascript tests pass with protocol `3.0.0-1`. 
- streamr-broker: tests pass with `3.0.0-1`

I then ran a broker node with protocol `3.0.0-1` along with the rest of the docker stack.
- streamr-client-javascript: tests pass both with protocol version `3.0.0-1` and `2.2.15` (current).
- streamr-client-java: tests pass (master branch)
- streamr-platform: tests pass (development branch)

I think based on this testing we can move along with the version upgrade?

mirpo commented:
LGFM
